Document how to use secrets with `if:` conditionals in GitHub Actions workflows

br3ndonland opened the pull request:
### Why:

Closes github/docs#6861

Users have requested documentation on how to use secrets with `if:` conditionals, especially how to skip a GitHub Actions workflow step if a secret isn't present.

This PR is based on the helpful guidance provided in https://github.com/github/docs/issues/6861#issuecomment-870757186.

Related to this, as far as I know, the docs don't mention the default return value for an expression referencing an unset secret. Based on trial and error, the return value of `${{ secrets.SECRET_IS_NOT_SET }}` appears to be an empty string. I've been digging around [actions/runner](https://github.com/actions/runner) to try to find where this default value is set, but I'm not sure. Maybe in [`SecretUtility.cs`](https://github.com/actions/runner/blob/HEAD/src/Sdk/Common/Common/Utility/SecretUtility.cs)?

### What's being changed:

- Add a complete workflow example to [`jobs.<job_id>.steps[*].if`](https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_idstepsif), demonstrating how to skip a step if a secret is not present
- Add an explanation to "[Using encrypted secrets in a workflow](https://docs.github.com/en/actions/security-guides/encrypted-secrets#using-encrypted-secrets-in-a-workflow)"
- Cross-reference the two pages

### Check off the following:

- [x] I have reviewed my changes in staging (look for "Automatically generated comment" and click **Modified** to view your latest changes).
- [x] For content changes, I have completed the [self-review checklist](https://github.com/github/docs/blob/main/contributing/self-review.md#self-review).

### Writer impact (This section is for GitHub staff members only):

- [ ] This pull request impacts the contribution experience
  - [ ] I have added the 'writer impact' label
  - [ ] I have added a description and/or a video demo of the changes below (e.g. a "before and after video")

<!-- Description of the writer impact here -->

br3ndonland committed:
:lock: Document how to use secrets with `if:`

github/docs#6861
github/docs#12722

- Add a complete workflow example to `jobs.<job_id>.steps[*].if`,
  demonstrating how to skip a step if a secret is not present
- Add an explanation to "Using encrypted secrets in a workflow"
- Cross-reference the two pages

ramyaparimi commented:
@br3ndonland 
Thanks so much for opening a PR! I'll get this triaged for review :zap:

danielray1 reviewed (commented):
I approve these changes on which I do believe it's rite an will work as it should

danielray1 reviewed (commented):
I do think this is the best way to handle it

Dorota1661 commented:
❤️

sob., 11 gru 2021, 00:16 użytkownik danielray1 ***@***.***>
napisał:

> ***@***.**** commented on this pull request.
>
> I do think this is the best way to handle it
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/github/docs/pull/12722#pullrequestreview-829400491>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AWZFQ4ASRCH3BN7XGIQCDHDUQKC37ANCNFSM5JSL6HIA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

Dorota1661 commented:
Dziękuję

Awan27091987 commented:
All delete device secre all

Dorota1661 commented:
1772

Dorota1661 commented:
12772

skedwards88 commented:
@br3ndonland Thank you for this addition! I think this info will be super helpful to users. I don't have permissions to edit your PR to resolve the merge conflict. The file has been renamed to `content/actions/using-workflows/workflow-syntax-for-github-actions.md`, so if you move your addition to that file, I think that should resolve the conflict. Once you resolve the conflict, we'll get this reviewed and merged for you.

ethomson commented on content/actions/using-workflows/workflow-syntax-for-github-actions.md:
```diff
@@ -342,6 +342,34 @@ steps:
+      SECRET_IS_SET: ${{ secrets.SECRET_IS_SET }}
+      SECRET_IS_NOT_SET: ${{ secrets.SECRET_IS_NOT_SET }}
+    steps:
+      - if: ${{ env.SECRET_IS_SET }}
```
This will be truthy, but I wonder if an explicit comparison to an empty string would really drive the point home?

br3ndonland replied:
Good suggestion @ethomson, thank you. -> aa53817

Rather than referencing two secrets (`${{ secrets.SECRET_IS_SET }}` and `${{ secrets.SECRET_IS_NOT_SET }}`), I rewrote this section to reference a single secret (`${{ secrets.SECRET_IS_SET }}`), and updated the `if:` conditionals to compare with empty strings as suggested. Hopefully this is a little clearer.

br3ndonland replied:
Updated again to use same environment variable and secret names as the cross-referenced "Encrypted secrets" page (e0b2cfb).

br3ndonland committed:
:lock: Compare secrets with empty strings in `if:`

github/docs#6861
https://github.com/github/docs/pull/12722#discussion_r801011000

Rather than referencing two secrets:

1. `${{ secrets.SECRET_IS_SET }}`
2. `${{ secrets.SECRET_IS_NOT_SET }}`)

This commit will update the related section of the docs to reference a
single secret (`${{ secrets.SECRET_IS_SET }}`), and will update the
`if:` conditionals to compare with empty strings as suggested.

br3ndonland committed:
:lock: Add missing `{% raw %}`/`{% endraw %}`

github/docs#6861
github/docs#12722

Some `${{ }}` values were converted to `$` in the preview environment.
Adding `{% raw %}`/`{% endraw %}` will preserve the raw value.

br3ndonland committed:
:lock: Match variable and secret names in examples

github/docs#6861
https://github.com/github/docs/pull/12722#discussion_r801011000

This PR adds an example of how to use secrets with `if:` conditionals.
The reviewer suggested comparing variable values with empty strings to
make the `if:` conditionals clearer.

Commit cecdf00 updated the secret names accordingly, but the names of
the secret and environment variable may still have been confusing.

This commit will update the secret and environment variable names to
match the cross-referenced example on the "Encrypted secrets" page.

hubwriter committed:
Merge branch 'main' into if-secrets

hubwriter commented on content/actions/using-workflows/workflow-syntax-for-github-actions.md:
````diff
@@ -342,6 +342,31 @@ steps:
     uses: actions/heroku@1.0.0
 ```
 
+### Example: Using secrets
````
To match level of previous subheadings:

```suggestion
#### Example: Using secrets
```

hubwriter committed:
Update content/actions/using-workflows/workflow-syntax-for-github-actions.md

hubwriter reviewed (approved):
@br3ndonland - This looks great. Thanks for making this change. 

The example is really good. I'm sure this will help a lot of people. 👍 

I'll get this merged/published.

ramyaparimi committed:
Merge branch 'main' into if-secrets

ramyaparimi committed:
Merge branch 'main' into if-secrets

m-kuhn commented:
I think this would deserve a hint on the implication that the secret is leaked to every step within the job, regardless if it's required for a certain step or not (hence violating the principle of least privilege).
